### ISSUE TITLE
AOT compiled function inside JIT function fails.

### ISSUE THREAD

**Issue by thijsvanputten** [numba/numba#5910]
I am trying to use numba to speed up a project, but I have hit an issue trying to use an AOT-compiled function inside a JIT-compiled function. Is this expected to be supported?

My reason for wanting to do this is that I have a family of physical functions that are very slow to compile and that are pretty much set in stone, with a caller function that is fast to compile and likely to change, which made me want to compile the first set AOT and the caller JIT.

I've made a minimal working example with the following structure:
```
test
-- __init__.py
-- foo_src.py
-- test.py`
```

foo_src.py contains:
```
from numba import jit
from numba.pycc import CC

cc = CC('foo')

@jit(nopython=True)
@cc.export('bar', 'i8(i8)')
def bar(x):
    return x * 4

if __name__ == "__main__":
    cc.compile()`
```

test.py contains:
```
from numba import jit
import foo

@jit(nopython=True)
def baz(x):
    y = 0
    for i in range(x):
        y += foo.bar(i)
    return y

if __name__ == "__main__":
    print (baz(10))`
```
The error message I get is:
```
Unknown attribute 'bar' of type Module(<module 'foo' from 'D:\\pythontest\\foo.cp37-win_amd64.pyd'>)
```

I've checked that it works fine with the same functions if make both JIT or both AOT, but the combination seems to fail.

**Comment by HPLegion** [comment]
I am not very familiar with the AOT features, but maybe you can benefit from `jit(cache=True)` in your situation? This way you would keep the flexibility of jit compilation and still not have to recompile the functions every time you start a new interpreter session.

**Comment by esc** [comment]
@thijsvanputten thanks for submitting this. Perhaps I missed it, but would you be so kind as to outline which commands to execute to reproduce the error?

**Comment by thijsvanputten** [comment]
> @thijsvanputten thanks for submitting this. Perhaps I missed it, but would you be so kind as to outline which commands to execute to reproduce the error?

Sorry, I simply run (from inside the test directory):
```
python foo_src.py
python test.py
```

**Comment by esc** [comment]
@thijsvanputten thanks for getting back to me, I can confirm that I am able to reproduce the issue.

**Comment by thijsvanputten** [comment]
> I am not very familiar with the AOT features, but maybe you can benefit from `jit(cache=True)` in your situation? This way you would keep the flexibility of jit compilation and still not have to recompile the functions every time you start a new interpreter session.

This is an option indeed. However, I'd prefer to freeze the lower level functions in compiled form for myself and my collaborators once, instead of having to rely on a local cache.
I also feel like with the cached option the compilation of the caller jit-function is still much slower than it needs to be, perhaps because it's inlining the lower level functions?

**Comment by esc** [comment]
@thijsvanputten thanks, I have updated this to be a feature request, as I am not sure there is a way to do what you want right now.

**Comment by KCrux** [comment]
I have exactly the same usecase as @thijsvanputten has. This would a really great feature!

**Comment by amagenPANW** [comment]
Joining the feature request, this will be great!

**Comment by BrandonBahret** [comment]
Would the fact that AOT generates platform-specific machine code be a hurdle here? At least, it would clash with an ideal of JITing being a means of avoiding shipping with binaries.

**Comment by stuartarchibald** [comment]
> Would the fact that AOT generates platform-specific machine code be a hurdle here? At least, it would clash with an ideal of JITing being a means of avoiding shipping with binaries.

Not necessarily, AOT compiled code targetting a basic instruction set for a given architecture should be callable from JIT functions. This is similar to e.g. using a `ctypes` binding to a system library and calling that from a JIT function.

IIRC the reason calling AOT compiled code from JIT compiled fuctions doesn't currently work is simply that the AOT module wraps the compiled functions in a "CPython wrapper" which permits calling them directly from Python, but prevents them from being called from JIT code.

**Comment by bairdlampard534** [comment]
@stuartarchibald could you give us a pointer to how to solve this issue? can `ctypes` be used here?

**Comment by stuartarchibald** [comment]
> @stuartarchibald could you give us a pointer to how to solve this issue? can `ctypes` be used here?

@bairdlampard534 I think it would requires changes to the way `pycc` works. The symbols for the exported functions are not global (external) in the DSO (shared library) that is produced by `pycc`. If you run `nm` on any DSO produced by `pycc`, you'll note that the `cc.export` compiled functions are in the text section but local, which means there's nothing to easily available against which to bind.

**Comment by bairdlampard534** [comment]
Thanks for the insight. What would be the fundamental difference between the output of caching vs. AOT compilation?  Caching generates `nbi` and `nbc` files. Can AOT compilation produce something similar and be loaded as compiled jit functions?

**Comment by stuartarchibald** [comment]
> Thanks for the insight. What would be the fundamental difference between the output of caching vs. AOT compilation? Caching generates `nbi` and `nbc` files. Can AOT compilation produce something similar and be loaded as compiled jit functions?

Fundamentally a Numba AOT library is a Python module compiled via the Python C-extension mechanism. The functions exposed in the AOT library are functions that can be called from the Python interpreter. For each of these functions Numba's AOT compiler has generated compiled-in call wrappers to convert from the Python representation of objects to the Numba representation, and then more code to deal with converting to Numba's calling convention, at which point something like the underlying function that would be compiled with `@njit` can be called. There's no technical reason why the Numba JIT function that is being called from the "Python" function couldn't be exported in the DSO, however, to call it via that symbol, the Numba calling convention would have to be used (it's quite involved as Numba has it's own state representation for things like exceptions and the data model of all the arguments is "flattened", e.g. a NumPy array representation is something like 6 separate arguments).

A function that is "cached" in the Numba cache is essentially the result of a `reduce()` call on a `CompileResult`: https://github.com/numba/numba/blob/1f1cee2d06d56633874f8c3a7df24245c6f26bd1/numba/core/compiler.py#L193-L207
most of the size of the cached object will be the serialized library, which is basically just storing the ELF bytes of the function.

Numba's cache isn't really designed to be distributed, it relies on pickle, which is not intended for long term storage, and it also doesn't have a specification or versioning. So it isn't really a viable target for AOT compilation. There's also the issue that JIT-cached functions are going to contain instructions specific to the CPU on which they were JIT compiled, which means they aren't necessarily portable for the purposes of redistribution either.
